### PR TITLE
Added optimize kwarg to torch.einsum

### PR DESCRIPTION
**Pull request by heitorschueroff**
This PR adds the `optimize` keyword argument to `torch.einsum` which accepts a contraction path in the same format given by `opt_einsum.contract_path()`. In the future, this keyword argument may be extended to accept more input types to be compatible with NumPy's `einsum`.

TODO

- [x] Fix ONNX failure
- [ ] Run some benchmarks

### CODE REVIEW

File: aten/src/ATen/native/Linear.cpp [hunk 1]
```diff
@@ -395,87 +416,100 @@ Tensor einsum(std::string equation, TensorList operands) {
...
-  // Fast path for when an operand has zero sized dim
-  if (has_zero_size_dim) {
-    std::vector<int64_t> out_shape(out_size);
-    for (const auto i : c10::irange(out_size)) {
-      out_shape[i] = permuted_operands[dim_last_op[i]].size(i);
-    }
-    return at::zeros(out_shape, result.options());
-  }
+  while (permuted_operands.size() > 1) {
+    decltype(permuted_operands.size()) i = 0, j = 1;
```
Comment (mruberry) [inline aten/src/ATen/native/Linear.cpp:455]: one variable construction per line, also it's probably easier to read:

```
auto i = decltype(permuted_operands.size()){0};
```

File: aten/src/ATen/native/Linear.cpp [hunk 2]
```diff
@@ -395,87 +416,100 @@ Tensor einsum(std::string equation, TensorList operands) {
...
+    permuted_operands.erase(permuted_operands.begin() + i);
+
+    if (has_zero_size_dim) {
+      permuted_operands.emplace_back(a.mul(b));
+      continue;
+    }
+
+    // Contract dimensions for which no other operands have size != 1
+    std::vector<int64_t> sum_dims;
+    for (int64_t dim = out_size; dim < perm_index; ++dim) {
```
Comment (mruberry) [inline aten/src/ATen/native/Linear.cpp:487]: `auto dim`?

File: torch/functional.py [hunk 3]
```diff
@@ -404,7 +404,17 @@ def einsum(equation, *operands):
         _operands = operands[0]
         # recurse incase operands contains value that has torch function
         # in the original implementation this line is omitted
-        return einsum(equation, *_operands)
+        return einsum(equation, *_operands, optimize=optimize)
+
+    if optimize is not None:
```
Comment (mruberry) [inline torch/functional.py:409]: Add a comment here

File: torch/functional.py [hunk 4]
```diff
@@ -293,7 +293,7 @@ def lu_unpack(LU_data, LU_pivots, unpack_data=True, unpack_pivots=True):
     return P, L, U
 
 
-def einsum(equation, *operands):
+def einsum(equation, *operands, optimize=None):
```
Comment (mruberry) [inline torch/functional.py:296]: Docs update explaining the kwarg?

**Comment by mruberry** [comment]
I'm looking forward to seeing the doc updates describing the kwarg!

ONNX test failure is real:

```
Apr 20 20:55:14 =================================== FAILURES ===================================
Apr 20 20:55:14 _____________________ TestONNXRuntime_opset12.test_einsum ______________________
```

XLA failure is real, too:

```
Apr 20 21:25:48 [  FAILED  ] AtenXlaTensorTest.TestEinsumPyTorchLowerRepeatedAxis
```

We might need to coordinate an update with @JackCaoG, or design a workaround so XLA doesn't have to update or support this kwarg. @JackCaoG, what's your thinking on adding an "optimize" kwarg to einsum? Is there some way we can avoid bothering you with this for now?

File: test/test_linalg.py [hunk 5]
```diff
@@ -4303,8 +4308,10 @@ def check(equation, *operands):
     @dtypes(torch.double, torch.cdouble)
     def test_einsum_random(self, device, dtype):
         def check(equation, *operands):
-            ref = np.einsum(equation, *[op.cpu().numpy() for op in operands])
-            res = torch.einsum(equation, operands)
+            np_operands = [op.cpu().numpy() for op in operands]
+            path = np.einsum_path(equation, *np_operands)[0]
```
Comment (mruberry) [inline test/test_linalg.py:4312]: Add a comment here. 

Should this be testing einsum vs. einsum (without the optimize kwarg) and then einsum_path vs einsum (with the optimize kwarg)?

File: test/test_linalg.py [hunk 6]
```diff
@@ -4226,10 +4226,15 @@ def test_qr_error_cases(self, device, dtype):
     @dtypes(torch.double, torch.cdouble)
     def test_einsum(self, device, dtype):
         def check(equation, *operands):
-            ref = np.einsum(equation, *[operand.cpu().numpy() for operand in operands])
+            np_operands = [operand.cpu().numpy() for operand in operands]
```
Comment (mruberry) [inline test/test_linalg.py:4229]: Similar question about test coverage here as below

**Comment by JackCaoG** [comment]
@mruberry The xla failure actually coming from the equation `"ijj,k->ik"` no longer uses `xla::Mul` and xla test was checking if `xla::Mul` being called to verify that `torch::einsum` correctly uses the xla ops. 

I can make the test to check `xla::bmm` for now, this is not a very good way to verify einsum but I don't have a better way right now.

File: torch/functional.py (hunk 4 above)
Comment (heitorschueroff) [reply torch/functional.py:296]: I added the Keyword Args: section

**Comment by heitorschueroff** [comment]
> @mruberry The xla failure actually coming from the equation `"ijj,k->ik"` no longer uses `xla::Mul` and xla test was checking if `xla::Mul` being called to verify that `torch::einsum` correctly uses the xla ops.
> 
> I can make the test to check `xla::bmm` for now, this is not a very good way to verify einsum but I don't have a better way right now.

@JackCaoG If my understanding is correct, this is because I removed the special cases when contracting the tensors and I'm now always calling into sumproduct_pair helper function? If so, I can add those special cases back.

**Comment by JackCaoG** [comment]
@heitorschueroff Nah, I think we should not change the code to fit a (poorly written) test, I submitted pr https://github.com/pytorch/xla/pull/2908 to update the test. There seems to be some build issue through, do you mind rebase this pr? Thanks!

**Comment by mruberry** [comment]
> @mruberry The xla failure actually coming from the equation `"ijj,k->ik"` no longer uses `xla::Mul` and xla test was checking if `xla::Mul` being called to verify that `torch::einsum` correctly uses the xla ops.
> 
> I can make the test to check `xla::bmm` for now, this is not a very good way to verify einsum but I don't have a better way right now.

Thanks @JackCaoG! You're the best; when are you going to ask us for a favor?

**Comment by JackCaoG** [comment]
> > @mruberry The xla failure actually coming from the equation `"ijj,k->ik"` no longer uses `xla::Mul` and xla test was checking if `xla::Mul` being called to verify that `torch::einsum` correctly uses the xla ops.
> > I can make the test to check `xla::bmm` for now, this is not a very good way to verify einsum but I don't have a better way right now.
> 
> Thanks @JackCaoG! You're the best; when are you going to ask us for a favor?

😛 Really looking forward to attend the pytorch dev day in the Europe(when they happen, if that happen).

**Comment by heitorschueroff** [comment]
closing this in favor of https://github.com/pytorch/pytorch/pull/60191

**Pull request closed without merging**